test/libsolidity/SyntaxTest.cpp

```diff
@@ -59,93 +59,90 @@ SyntaxTest::SyntaxTest(string const& _filename)
+		{
+			solAssert(currentError->comment(), "");
+			if (currentError->type() == Error::Type::Warning)
+			{
+				bool ignoreWarning = false;
+				for (auto const& filter: m_warningsToFilter)
+					if (currentError->comment()->find(filter) == 0)
+					{
+						ignoreWarning = true;
+						break;
+					}
+				if (ignoreWarning)
+					continue;
+			}
+
+			m_errorList.emplace_back(SyntaxTestError{currentError->typeName(), errorMessage(*currentError)});
+		}
+	}
+	catch (CompilerError const& _e)
+	{
+		m_errorList.emplace_back(SyntaxTestError{"CompilerError", errorMessage(_e)});
+	}
+	catch (InternalCompilerError const& _e)
+	{
+		m_errorList.emplace_back(SyntaxTestError{"InternalCompilerError", errorMessage(_e)});
+	}
+	catch (FatalError const& _e)
+	{
+		m_errorList.emplace_back(SyntaxTestError{"FatalError", errorMessage(_e)});
+	}
+	catch (UnimplementedFeatureError const& _e)
+	{
+		m_errorList.emplace_back(SyntaxTestError{"UnimplementedFeatureError", errorMessage(_e)});
+	}
+	catch (Exception const& _e)
+	{
+		m_errorList.emplace_back(SyntaxTestError{"Unknown Exception", errorMessage(_e)});
+	}
+
+	if (m_expectations != m_errorList)
 	{
 		std::string nextIndentLevel = _linePrefix + "  ";
 		FormattedScope(_stream, _formatted, {BOLD, CYAN}) << _linePrefix << "Expected result:" << endl;
-		printExpected(_stream, nextIndentLevel, _formatted);
+		printErrorList(_stream, m_expectations, nextIndentLevel, _formatted);
 		FormattedScope(_stream, _formatted, {BOLD, CYAN}) << _linePrefix << "Obtained result:\n";
-		printErrorList(_stream, m_errorList, nextIndentLevel, false, false, _formatted);
+		printErrorList(_stream, m_errorList, nextIndentLevel, _formatted);
 		return false;
 	}
 	return true;
 }
 
-void SyntaxTest::printExpected(ostream& _stream, string const& _linePrefix, bool const _formatted) const
-{
-	if (m_expectations.empty())
-		FormattedScope(_stream, _formatted, {BOLD, GREEN}) << _linePrefix << "Success" << endl;
-	else
-		for (auto const& expectation: m_expectations)
-		{
-			FormattedScope(_stream, _formatted, {BOLD, expectation.type == "Warning" ? YELLOW : RED}) <<
-				_linePrefix << expectation.type << ": ";
-			_stream << expectation.message << endl;
-		}
-}
-
 void SyntaxTest::printErrorList(
 	ostream& _stream,
-	ErrorList const& _errorList,
+	std::vector<SyntaxTestError> const& _errorList,
 	string const& _linePrefix,
-	bool const _ignoreWarnings,
-	bool const _lineNumbers,
 	bool const _formatted
-) const
+)
 {
 	if (_errorList.empty())
 		FormattedScope(_stream, _formatted, {BOLD, GREEN}) << _linePrefix << "Success" << endl;
 	else
 		for (auto const& error: _errorList)
 		{
-			bool isWarning = (error->type() == Error::Type::Warning);
-			if (isWarning && _ignoreWarnings) continue;
-
 			{
-				FormattedScope scope(_stream, _formatted, {BOLD, isWarning ? YELLOW : RED});
+				FormattedScope scope(_stream, _formatted, {BOLD, (error.type == "Warning") ? YELLOW : RED});
 				_stream << _linePrefix;
-				if (_lineNumbers)
```

Comment:
Was this removed on purpose?

Reply:
Yes, this was removed on purpose.

Before this pull request line numbers were only displayed for parser errors - the idea being that those errors should never occur and have to be fixed to construct a valid test contract. For syntax errors no line numbers were displayed, because they are not included in the definition of test expectations and displaying them might be confusing.

Now parser tests are valid test cases as well, so for the same reason I removed displaying line numbers for parser errors as well.

Reply:
It might be helpful to include line numbers in the expectations, because some expectations might succeed accidentally otherwise.

Reply:
Oh but then it might actually be better to just include the unconverted start-end-indices, so that we also see when the source location is wrong.

Reply:
OK - that should probably be a separate pull request then, though.

Reply:
I created a separate issue for this: #3810.